### ISSUE TITLE
Custom layout issue 

### ISSUE THREAD

**Issue by cderv** [quarto-dev/quarto-cli#8187]
Example from https://quarto.org/docs/authoring/figures.html#custom-layout not working anymore 

````markdown
---
format: 
  html: default
  pdf: default
keep-md: true
---

```{r}
#| layout: [[45,-10, 45], [100]]

plot(cars)
plot(pressure)
plot(mtcars)
```

````

This is not correct  layout. 

<details>
<summary>Intermediate md</summary>

````markdown
---
format: 
  html: default
  pdf: default
keep-md: true
---


::: {.cell layout="[[45,-10,45],[100]]"}

```{.r .cell-code}
plot(cars)
```

::: {.cell-output-display}
![](index_files/figure-html/unnamed-chunk-1-1.png){width=672}
:::

```{.r .cell-code}
plot(pressure)
```

::: {.cell-output-display}
![](index_files/figure-html/unnamed-chunk-1-2.png){width=672}
:::

```{.r .cell-code}
plot(mtcars)
```

::: {.cell-output-display}
![](index_files/figure-html/unnamed-chunk-1-3.png){width=672}
:::
:::

````

</details>

![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/82f83589-3ddc-44f1-8f1b-9d76b78e776a)

**Comment by cscheid** [comment]
Part of the problem is that we're only lifting code cells out of _floats_ with layouts, and not out of standalone layouts. (That's a bug)

**Comment by cscheid** [comment]
Consider this:


````
---
title: foo
keep-md: true
---

```{r}
#| label: fig-1
#| layout: [[45,-10, 45], [100]]
#| fig-cap: Hello
#| fig-subcap: true

plot(cars)
plot(pressure)
plot(mtcars)
```
````

<img width="804" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/285675/e9813f8c-1f63-4f0e-9b9c-961b0673f574">

**Comment by cderv** [comment]
I was doing a git bisect to help pinpoint the change. The breaking happened in 361993dd8ee7b9ea5e0610706f63a8eaf108c6be as part of https://github.com/quarto-dev/quarto-cli/pull/7976 to solve https://github.com/quarto-dev/quarto-cli/issues/6157

Though we don't have cell annotation here, so maybe something related to more scoped fix for code annotation thing. I did not look yet at the change

**Comment by cderv** [comment]
I understand the issue is that the filter moved in #7976 is applying on three `CodeBlock` instead of one because it has not been merged into one yet. 

https://github.com/quarto-dev/quarto-cli/blob/9f6968358fb00034c08cc613294c906a184a853a/src/resources/filters/main.lua#L391-L392

`fold_code_and_lift_codeblocks()` is applied on `Codeblock` before `PanelLayout` is created with only one CodeBlock merged. 

The change in order created  this regression because this is applied now on each CodeBlock
https://github.com/quarto-dev/quarto-cli/blob/9f6968358fb00034c08cc613294c906a184a853a/src/resources/filters/quarto-post/foldcode.lua#L123-L125
which applies `make_code_cell_scaffold`
https://github.com/quarto-dev/quarto-cli/blob/9f6968358fb00034c08cc613294c906a184a853a/src/resources/filters/quarto-post/foldcode.lua#L46-L56

which probably messes up the layout code after... 

I really don't master nor comprehend yet this scaffold thing so that is what I am writing that for future reference to document my learning and understanding of behavior. 

If it can help know if we do the right things, that is additional benefit!

**Comment by cscheid** [comment]
> this scaffold thing

A "scaffold" in the Lua filters is just a div that we add so that the contents are inside a single Div during the processing, but removed before the final rendering (because some formats don't like it when there's nested Divs like that). see `descaffold.lua`

**Comment by cscheid** [comment]
If we flip the order back, then code-fold will break again. I think that the right fix is for the code-merging filter to work on any layouts, not just FloatRefTarget ones.

**Comment by cscheid** [comment]
Ok, one fix for the immediate issue is, indeed, to simply not emit a scaffolding as we're processing `code-fold` when the cells don't change. Let's see what the test suite says.

**Comment by cderv** [comment]
> A "scaffold" in the Lua filters is just a div that we add so that the contents are inside a single Div during the processing, but removed before the final rendering (because some formats don't like it when there's nested Divs like that). see descaffold.lua

Ok thanks. I'll be curious to understand why we are doing that - new stuff to me. When you have time. 

> If we flip the order back, then code-fold will break again.

By flipping the order, we indeed made the CodeBlock scaffolded and it seems layout does not account for that. Anyhow, I get it better now and you'll know better. thanks !

**Comment by cscheid** [comment]
> I'll be curious to understand why we are doing that

It can make our processing simpler, because we know that the element type will always be a Div. Here, I had added it so that the downstream code didn't need to check whether the node itself was a Div, a CodeBlock, or something else. By making a scaffolding, it was clearer that we always had a Div, and so we could always insert elements in `div.content`.

In this specific case, this technique failed because some _other_ downstream code was iterating over nodes directly rather than calling `walk()`, and so the change in structure caused the merging to fail. See [the PR](https://github.com/quarto-dev/quarto-cli/pull/8189) for how I fixed it.
